### PR TITLE
bindable action buttons

### PR DESCRIPTION
**Pull request by mc-oofert**
### work started 12/12/2023

you may alt-click action buttons to bind them to a key
these are subject to click cooldown and if an action successfully triggers click cooldown is triggered so you cant instantly do multiple
alt-click again to unbind
## Why It's Good For The Game

moving your mouse to the top left corner to do combat is not good gameplay

## Changelog
:cl:
qol: you may altclick action buttons to bind them to a key
/:cl:

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
Closing this for breaching feature freeze.

**Comment by JohnFulpWillard** [comment]
They started working on it before the freeze

**Comment by mc-oofert** [comment]
draft until i figure out how to fix linters

**Comment by mc-oofert** [comment]
actually im drafting this before i go to sleep because things need to be done

File: code/modules/mob/mob_defines.dm [hunk 1]
```diff
@@ -200,4 +200,6 @@
 	var/active_typing_indicator
 	///the icon currently used for the thinking indicator's bubble
 	var/active_thinking_indicator
-
+	
+	/// key pressed to weakref of an action
+	var/key_to_action_weakref = list()
```
[the remark this answers is not in the document]
Comment (mc-oofert) [reply code/modules/mob/mob_defines.dm]: idk how the profiler works though

**Comment by mc-oofert** [comment]
i was told my implementation sucks

**Comment by Eronymun** [comment]
I can't believe no has commented on this. This is probably one of the greatest UX changes being worked on right now - I've been asking for bind-able action keys for years. Awesome that you picked it up.

File: code/modules/tgui_input/keycombo.dm [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ */
+/proc/tgui_input_keycombo(mob/user = usr, message, title = "Key Input", default = 0, timeout = 0, ui_state = GLOB.always_state)
+	if (!istype(user))
+		if (istype(user, /client))
+			var/client/client = user
+			user = client.mob
+		else
+			return null
+
+	if (isnull(user.client))
```
Comment (Ghommie) [inline code/modules/tgui_input/keycombo.dm:21]:
```suggestion
	if (!istype(user))
		if (!istype(user, /client))
			return null
		var/client/client = user
		user = client.mob
	else if (isnull(user.client))
```
clients cannot be mobless afaik, so skipping the null client check if the user arg was a client before being changed to mob should be fine.

**Review by Ghommie — APPROVED** [review]

**Comment by mc-oofert** [comment]
This should be testmerged

**Comment by mc-oofert** [comment]
No issues seem to have been reported

**Pull request merged**